tensorpipe/channel/channel.cc

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/channel.h>
+#include <tensorpipe/common/tensor.h>
+
+namespace tensorpipe {
+namespace channel {
+template <>
+void Channel<CpuTensor>::send(
+    const void* ptr,
+    size_t length,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  send(
+      CpuTensor{const_cast<void*>(ptr), length},
```

Comment:
I see why we need the const_cast here, but it looks like a code smell. In other words, by introducing these CpuTensor structs, are we giving up some const-correctness? Because a const CpuTensor means that the address of its pointers is const, but it doesn't enforce constness on the memory they point to. I'm wondering if there is some semi-standard way to deal with this? If not, well, too bad, let's go ahead anyways...

Reply:
As we agreed we don't need backwards compatibility for the `Channel` interface, this `const_cast` itself disappeared, but the point remains: what previously was a `const void*` parameter is now a `void*` member of a struct. We could make the corresponding member of `CpuTensor` a `const void*` and introduce a `MutableCpuTensor` (or the other way around), but I'm not sure the benefits outweigh the cost.
An other option would be to introduce getters  in `CpuTensor` (i.e. `void* getPtr()` and `const void* getPtr() const`). What do you think?

Reply:
As we agreed we don't need backwards compatibility for the `Channel` interface, this `const_cast` itself disappeared, but the point remains: what previously was a `const void*` parameter is now a `void*` member of a struct. We could make the corresponding member of `CpuTensor` a `const void*` and introduce a `MutableCpuTensor` (or the other way around), but I'm not sure the benefits outweigh the cost.
An other option would be to introduce getters  in `CpuTensor` (i.e. `void* getPtr()` and `const void* getPtr() const`). What do you think?

Reply:
Using getters was indeed what I was thinking about... It may make the code a bit less readable though, not sure how that compares with the added static safety...

This is not a decision that directly affects the pipe API right? There, a Message is always passed by value (or moved, so by rvalue ref), hence it's never const. The const-vs-non-const choice only affects the channel's API. However, the Tensor's getters would then also be visible to the user of the pipe (because they are included in the Message) and they may look weird there. Is this reason enough to not introduce getters and give up the const-safety?